### DISCUSSION TITLE
`parseNumber` and `parseBoolean`

### DISCUSSION THREAD

**Discussion by jmcdo29** [jmcdo29/nest-commander#12] [Ideas]
How would the community feel about adding in methods for `parseBoolean` and `parseNumber` so that they didn't have to write them themselves? 

The `parseBoolean()` could go so far as to accept `yes`, `y`, `1`, `t`, and `true` for true values, and `no`, `n`, `0`, `f` and `false` for false values.

Anyone have any thoughts about it?

**Comment by smolinari** [comment]
As we were discussing, having/ using inquirer for input would be really cool. Also a coloring utility and ora too for enhancing output. Oh yeah! :)

Scott

**Comment by micalevisk** [comment]
that would be useful.

How would the interface looks like? They will belong to which class?

**Reply by jmcdo29** [reply]
I think the `parseBoolean` and `parseNumber` methods would just be exported methods. Not belonging to any particular class, though I suppose we _could_ make a `CLIUtilityService` if we really wanted to that has these. As for interfaces, I'd be thinking a simple
```ts
parseBoolean(val: string): boolean
parseNumber(val: string): number
```

**Reply by micalevisk** [reply]
<details>
<summary>v1: functions exported by <code>nest-commander</code></summary>

```ts
import { Command, CommandRunner, Option, parseNumber, parseBoolean } from 'nest-commander'

interface BasicCommandOptions {
  bool: boolean
  num: number
}

@Command({ name: 'basic' })
export class BasicCommand implements CommandRunner {
  async run(passedParam: string[], options: BasicCommandOptions) {}

  @Option({
    flags: '-n, --num <num>',
    required: true,
  })
  parseNum(val: string): number {
    return parseNumber(val)
  }

  @Option({
    flags: '-b, --bool <bool>',
    required: true,
  })
  parseBool(val: string): boolean {
    return parseBoolean(val)
  }
}
```

</details>

<details>
<summary>v2: under a new provider <code>CLIUtilityService</code></summary>

```ts
import { Command, CommandRunner, Option, CLIUtilityService } from 'nest-commander'

interface BasicCommandOptions {
  bool: boolean
  num: number
}

@Command({ name: 'basic' })
export class BasicCommand implements CommandRunner {
  constructor(private readonly util: CLIUtilityService) {}

  async run(passedParam: string[], options: BasicCommandOptions) {}

  @Option({
    flags: '-n, --num <num>',
    required: true,
  })
  parseNum(val: string): number {
    return this.util.parseNumber(val)
  }

  @Option({
    flags: '-b, --bool <bool>',
    required: true,
  })
  parseBool(val: string): boolean {
    return this.util.parseBoolean(val)
  }
}
```

</details>

---

To me, v2 feels better.

**Reply by jmcdo29** [reply]
V2 definitely feels more nest inspired. @smolinari you with with the package a good bit. Thoughts?

**Reply by micalevisk** [reply]
users could also extend `CLIUtilityService` class to, let's say, adding more parsers or overwrite some, like

```ts
class MyCLIUtilityService extends CLIUtilityService {
  parseNumber(val: string): 1 | 3 {
    const num = super.parseNumber(val)
    // ... throws if the value is not `1` nor `3`
  }
}
```

right?

**Reply by jmcdo29** [reply]
Right. They'll need a custom token for it and to add in the provider, but it would be doable

**Reply by micalevisk** [reply]
any cons on using `InquirerService`, and  
`parseNumber(val: string | undefined, opts?: { promptIfMissing:boolean }): number`  
`parseBoolean(val: string | undefined, opts?: { promptIfMissing:boolean }): boolean`

parsing + prompt might be pretty common. Other stuffs (like coloring and spinners) could be added later on under `opts` object

**Reply by micalevisk** [reply]
maybe `CLIParsersService` would be better for SoC principle

**Reply by smolinari** [reply]
I'd say, v2 is the better one. 

Scott

**Reply by jmcdo29** [reply]
Added #251 to address this
